File: test/fixtures/safe.go
// Idiomatic, safe Go. Any finding here is a false positive.
package main

import (
	"database/sql"
	"net/http"
	"os/exec"
)

func getUser(db *sql.DB, r *http.Request) (*sql.Rows, error) {
	// Placeholder — the driver binds the value.
	return db.Query("SELECT * FROM users WHERE id = $1", r.URL.Query().Get("id"))
}

func ping(r *http.Request) error {
	// Direct binary, separate arguments, no shell.
	host := r.FormValue("host")
	return exec.Command("ping", "-c", "1", host).Run()
}

File: test/fixtures/safe.php
<?php
// Idiomatic, safe PHP. Any finding here is a false positive.

function get_user(PDO $pdo) {
    // Prepared statement — the driver binds the value.
    $stmt = $pdo->prepare("SELECT * FROM users WHERE id = ?");
    $stmt->execute([$_GET['id']]);
    return $stmt->fetchAll();
}

function ping() {
    $host = $_POST['host'];
    $allowed = ['api.internal.example'];
    if (!in_array($host, $allowed, true)) {
        return null;
    }
    return shell_exec("ping -c 1 " . escapeshellarg($host));
}

function digest($data) {
    return hash('sha256', $data);
}

$password = getenv('DB_PASSWORD');
